src/hdbscan.jl

```diff
@@ -0,0 +1,303 @@
+                (c1, c2) = (c2, c1)
+            end
+            #record the lambda value
+            append!(clusters[c2].λp, fill(λ, length(clusters[c1])))
+            #unite cluster
+            unite!(uf, j, k)
+            #create parent cluster
+            points = members(uf, group(uf, j))
+            push!(clusters, HdbscanCluster(points))
+        else
+            #unite the noise cluster
+            unite!(uf, j, k)
+            #create parent cluster
+            points = members(uf, group(uf, j))
+            if length(points) < min_size
+                push!(clusters, HdbscanCluster(Int[]))
+            else
+                push!(clusters, HdbscanCluster(points))
+            end
+        end
+    end
+    @assert length(clusters) == 2n - 1
+    return clusters
+end
+
+function prune_cluster!(hierarchy::Vector{HdbscanCluster})
+    for i in 1 : length(hierarchy)-1
+        if isnoise(hierarchy[i])
+            c = hierarchy[i]
+            push!(hierarchy[c.parent].children, i)
+            hierarchy[c.parent].children_stability += c.stability
+        else
+            c = hierarchy[i]
+            if c.stability > c.children_stability
+                push!(hierarchy[c.parent].children, i)
+                hierarchy[c.parent].children_stability += c.stability
+            else
+                append!(hierarchy[c.parent].children, c.children)
+                hierarchy[c.parent].children_stability += c.children_stability
+            end
+        end
+    end
+end
+
+# Below are utility functions for building hierarchical trees
+heappush!(h, v) = insert!(h, searchsortedfirst(h, v), v)
+
+# Union-Find
+# structure for managing disjoint sets
+# This structure tracks which sets the elements of a set belong to,
+# and allows us to efficiently determine whether two elements belong to the same set.
+mutable struct UnionFind
+    parent:: Vector{Integer}  # parent[root] is the negative of the size
+    label::Dict{Int, Int}
+    next_id::Int
+
+    function UnionFind(nodes::Integer)
+        if nodes <= 0
+            throw(ArgumentError("invalid argument for nodes: $nodes"))
+        end
+
+        parent = -ones(nodes)
+        label = Dict([(i, i) for i in 1 : nodes])
+        new(parent, label, nodes)
+    end
+end
+
+# label of the set which element `x` belong to
+group(uf::UnionFind, x) = uf.label[root(uf, x)]
+# all elements that have the specified label
+members(uf::UnionFind, x::Int) = collect(keys(filter(n->n.second == x, uf.label)))
+
+# root of element `x`
+# The root is the representative element of the set
+function root(uf::UnionFind, x::Integer)
+    if uf.parent[x] < 0
+        return x
+    else
+        return uf.parent[x] = root(uf, uf.parent[x])
+    end
+end
+
+# whether element `x` and `y` belong to the same set
+function issame(uf::UnionFind, x::Integer, y::Integer)
+    return root(uf, x) == root(uf, y)
+end
+
+function Base.size(uf::UnionFind, x::Integer)
+    return -uf.parent[root(uf, x)]
+end
```

Comment:
That overload does not conform to `Base.size(obj, i)` API convention, which should give the size of `i`-th dimension of an array. This has to be absolutely avoided. Plus, we don't need to overload Base API for internal types.

It's not 100% clear what this function does, if it returns the set size, let's just call it this way:
```suggestion
function setsize(uf::UnionFind, x::Integer)
    return -uf.parent[root(uf, x)]
end
```